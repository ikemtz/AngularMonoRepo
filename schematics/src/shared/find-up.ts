import { dirname, parse, resolve } from "path";
import { lstatSync } from "fs";

export function findUpSync(name: string): string | undefined {
  let directory = __dirname;
  const { root } = parse(directory);

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const foundPath = locatePathSync(name, directory);
    if (foundPath) {
      return resolve(directory, foundPath);
    }
    else if (directory === root) {
      break;
    } else {
      directory = dirname(directory);
    }
  }
}

export function locatePathSync(
  file: string,
  directory: string
): string | undefined {
  try {
    const stat = lstatSync(resolve(directory, file));

    if (stat['isFile']) {
      return file;
    }
    // eslint-disable-next-line no-empty
  } catch { }
}