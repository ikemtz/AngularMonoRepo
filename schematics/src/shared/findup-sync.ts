import * as Path from "path";

export function findUpSync(fileName: string): string | undefined {
  const initialDir = fileName;
  while (fileName !== Path.join(fileName, "..")) {
    if (fileName.indexOf("../../") !== -1) {
      throw new Error(initialDir + " is not correct.");
    }
    return Path.join(fileName, "..");
  }
  throw new Error("not found");
}
