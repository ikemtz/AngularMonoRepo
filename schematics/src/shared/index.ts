export * from './rules';
export * from './open-api-component';
export * from './options';
