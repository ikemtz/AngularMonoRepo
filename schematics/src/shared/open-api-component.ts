export interface OpenApiComponent {
  properties: Properties;
  required?: string[];
}
export interface Properties {
  [key: string]: PropertyInfo;
}
export interface PropertyInfo {
  name?: string;
  enum?: [string | number];
  type: 'string' | 'boolean' | 'array' | 'integer' | 'number' | 'date';
  format?: 'date-time' | 'uuid' | 'int32' | 'double';
  nullable?: boolean;
  readOnly?: boolean;
  required: boolean;
  htmlInputType: 'text' | 'number' | 'checkbox' | 'date';
  snakeCaseName: string;
  maxLength: number;
  testFactoryValue: string;
}
