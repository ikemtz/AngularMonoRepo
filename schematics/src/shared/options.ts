import { PropertyInfo } from './open-api-component';

export interface IOptions {
  /** Entity Name (Must match schema name in json) */
  name: string;
  /** Parent Entity Name (Must match schema name in json) */
  parentName?: string;
  /** NGRX Store Feature Name */
  storeName?: string;
  /** Angular application Prefix */
  appPrefix: string;
  /** Folder to place generated files */
  path: string;
  /** Optional: Url of remote OpenApi JSON document (If this is not provided, then you must specify openApiJsonFileName) */
  openApiJsonUrl?: string;
  /** Optional: File path of local OpenApi JSON document (If this is not provided, then you must specify openApiJsonUrl) */
  openApiJsonFileName?: string;
  /** Optional: This is the property that will used for initial sorting on list and auto focus on crud schematics */
  modelFolderLocation?: string;
  /** Optional: Location of model files relative to the module/component generated root */

  firstProperty?: PropertyInfo;
  /** Ignore: This is for internal schematic use only  */
  swaggerProperties: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  swaggerObjectProperties?: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  hasDates?: boolean;
  hasObjects?: boolean;
  hasNullableDates?: boolean;
}
