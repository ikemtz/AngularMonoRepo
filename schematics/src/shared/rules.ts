import {
  apply,
  applyTemplates,
  chain,
  DirEntry,
  mergeWith,
  move,
  Rule,
  SchematicContext,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { strings, normalize } from '@angular-devkit/core';
import fetch from 'node-fetch';
import { PropertyInfo, OpenApiComponent } from './open-api-component';
import * as pluralize from 'pluralize';
import { Observable, from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { IOptions } from './options';
import _ = require('lodash');
import { TslintFixTask } from '@angular-devkit/schematics/tasks';
import * as fs from 'fs';
import * as findUp from 'find-up';
import * as https from 'https';
import * as http from 'http';

export function getSwaggerDoc(options: IOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    let jsonDoc: Observable<any> | null = null;
    if (options.openApiJsonFileName) {
      const fileName = getFileNames(options.openApiJsonFileName);
      context.logger.info(`Getting Swagger Document @${fileName}`);
      if (!fileName) {
        throw new Error(`File not found: ${options.openApiJsonFileName}`);
      }
      const response = fs.readFileSync(fileName);
      const apiDoc = JSON.parse(response.toString());
      jsonDoc = of(apiDoc);
    } else if (options.openApiJsonUrl) {
      context.logger.info(`Getting Swagger Document @${options.openApiJsonUrl}`);
      const httpsAgent = options.openApiJsonUrl.toLowerCase().startsWith("https") ? new https.Agent({
        rejectUnauthorized: false,
      }) : new http.Agent();
      jsonDoc = from(fetch(options.openApiJsonUrl, {
        agent: httpsAgent
      })).pipe(
        switchMap(resp => from(resp.json())));
    }
    if (jsonDoc) {
      return jsonDoc.pipe(map(data => processOpenApiDoc(data, options, host)));
    }
    return of(host);

  };
}

function getFileNames(openApiJsonFileName: string) {
  if (fs.existsSync(`${__dirname}/${openApiJsonFileName}`)) {
    return `${__dirname}/${openApiJsonFileName}`;
  }
  return findUp.sync(openApiJsonFileName);
}

export function processOpenApiDoc(data: any, schema: IOptions, host: Tree) {
  const openApiComonent = data.components.schemas[strings.classify(schema.name)] as OpenApiComponent;
  if (!openApiComonent) {
    throw new Error(`OpenApi Component not found in swagger doc: ${schema.name}`);
  }
  const properties = openApiComonent.properties as {
    [key: string]: PropertyInfo;
  };
  const filteredProperties: PropertyInfo[] = [];
  const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];
  schema.hasDates = false;
  for (const propertyKey in properties) {
    if (excludedFields.indexOf(propertyKey) < 0 && properties[propertyKey].type !== 'array') {
      const property =
        mapPropertyAttributes(schema, properties[propertyKey], {
          ...properties[propertyKey],
          name: propertyKey,
          required: (openApiComonent.required ?? []).indexOf(propertyKey) > -1,
        });
      filteredProperties.push(property);
    }
  }
  schema.swaggerProperties = filteredProperties;
  return host;
}

function mapPropertyAttributes(options: IOptions, source: PropertyInfo, dest: any) {

  if (source.type === 'number' || source.type === 'integer') {
    dest.htmlInputType = 'number';
    dest.filterExpression = 'numeric';
    dest.testFactoryValue = '0';
  } else if (source.type === 'boolean') {
    dest.htmlInputType = 'checkbox';
    dest.filterExpression = 'boolean';
    dest.testFactoryValue = 'true';
  } else if (source.format === 'date-time') {
    dest.htmlInputType = 'date';
    dest.filterExpression = 'date';
    dest.testFactoryValue = 'new Date()';
    options.hasDates = true;
  } else {
    dest.htmlInputType = 'text';
    dest.filterExpression = 'text';
    dest.testFactoryValue = `'${dest.maxLength ?
      _.snakeCase(dest.name).toUpperCase().substring(0, dest.maxLength) : _.snakeCase(dest.name).toUpperCase()}'`;
  }
  dest.snakeCaseName = _.snakeCase(dest.name);
  dest.startCaseName = _.startCase(dest.name);
  return dest;
}

export function generateFiles(schema: IOptions, templateType: 'list' | 'crud' | 'module'): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.logger.info(`Running ${strings.capitalize(templateType)} Schematic`);

    const singularizedName = pluralize(schema.name, 1);
    const pluralizedName = pluralize(schema.name, 2);
    const singularizedStoreName = schema.storeName ? pluralize(schema.storeName, 1) : '';
    const pluralizedStoreName = schema.storeName ? pluralize(schema.storeName, 2) : '';

    const folderName = normalize(`${schema.path}/${strings.dasherize(pluralizedName)}-${templateType}`);
    const templateSource = apply(url('./files'), [
      applyTemplates({
        ...strings,
        ...schema,
        pluralizedName,
        singularizedName,
        pluralizedStoreName,
        singularizedStoreName,
      }),
      move(folderName),
    ]);
    return chain([mergeWith(templateSource)])(tree, context);
  };
}

export function runLint(schema: IOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const dir: DirEntry | null = tree.getDir(schema.path.substr(0, schema.path.lastIndexOf('/')));
    const files = tree.actions.reduce((acc: Set<string>, action) => {
      const path = action.path.substr(1); // Remove the starting '/'.
      if (path.endsWith('.ts') && dir && action.path.startsWith(dir.path)) {
        acc.add(path);
      }
      return acc;
    }, new Set<string>());
    context.addTask(
      new TslintFixTask({
        ignoreErrors: false,
        tsConfigPath: 'tsconfig.json',
        files: Array.from(files),
      }),
    );
    return tree;
  };
}
