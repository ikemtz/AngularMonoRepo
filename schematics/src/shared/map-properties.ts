import * as pluralize from 'pluralize';
import { PropertyInfo, OpenApiComponent, OpenApiDocument } from './open-api-component';
import { IOptions } from './options';
import { mapPropertyAttributes } from './rules';

const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];

export function mapProperties(properties: { [key: string]: PropertyInfo; }, options: IOptions, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  let firstProperty: PropertyInfo | undefined;
  const filteredProperties: PropertyInfo[] = [];
  for (const propertyKey in properties) {
    const originalProperty = properties[propertyKey];
    if (excludedFields.indexOf(propertyKey) < 0 && originalProperty.type !== 'array') {
      const property = mapPropertyAttributes(options, originalProperty, {
        ...properties[propertyKey],
        name: propertyKey,
        required: (openApiComponent.required ?? []).indexOf(propertyKey) > -1,
      });

      mapReferencedProperties(property, propertyKey, openApiComponent, openApiDoc);
      if (!firstProperty && propertyKey !== 'id') {
        firstProperty = property;
      }
      filteredProperties.push(property);
    }
  }
  return { firstProperty, filteredProperties };
}
function mapReferencedProperties(property: PropertyInfo, propertyKey: string, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  const prop = openApiComponent.properties[propertyKey];
  const propTypeNames = prop['$ref']?.split('/');
  if (propTypeNames) {
    const typeName: string = propTypeNames.pop() || '';
    property.propertyTypeName = typeName;
    property.pluralizedPropertyTypeName = pluralize.plural(typeName);
    property.singularizedPropertyTypeName = pluralize.singular(typeName);
    const refComponent = openApiDoc.components.schemas[typeName];
    property.enum = refComponent.enum;
    if (property.enum) {
      const value = property.enum[0].toString().split(' ').pop();
      property.testFactoryValue = `${property.pluralizedPropertyTypeName}.${value}`
    }
  }
}

