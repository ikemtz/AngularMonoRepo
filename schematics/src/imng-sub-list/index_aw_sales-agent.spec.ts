import { Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { IOptions } from '../shared';

const collectionPath = path.join(__dirname, `../collection.json`);

describe(`imng-sub-list`, () => {
  let tree: UnitTestTree;
  let options: IOptions;
  beforeAll(async () => {
    const runner = new SchematicTestRunner(`schematics`, collectionPath);
    options = {
      name: `SalesAgent`,
      parentName: 'Customer',
      openApiJsonUrl: `https://awod-ikemtz.azurewebsites.net/swagger/v1/swagger.json`,
      path: `./test`,
      swaggerProperties: [],
      appPrefix: 'aw',
    };
    tree = await runner.runSchematic(`imng-sub-list`, options, Tree.empty());
  })
  test(`tree files should match`, () => {
    expect(tree.files).toMatchSnapshot();
  });
  test(`actions should work`, () => {
    const file = tree.get(`/test/+state/sales-agent.actions.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`effects should work`, () => {
    const file = tree.get(`/test/+state/sales-agent.effects.ts`
    );
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`selectors should work`, () => {
    const file = tree.get(`/test/+state/sales-agent.selectors.ts`
    );
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`models should work`, () => {
    const file = tree.get(`/test/models/ext-customer.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`api.service should work`, () => {
    const file = tree.get('/test/sales-agents-list/api.service.ts');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`list html should work`, () => {
    const file = tree.get('/test/sales-agents-list/list.component.html');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`list component spec should work`, () => {
    const file = tree.get('/test/sales-agents-list/list.component.spec.ts');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`list component should work`, () => {
    const file = tree.get('/test/sales-agents-list/list.component.ts');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`facade spec should work`, () => {
    const file = tree.get('/test/sales-agents-list/list.facade.spec.ts');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`facade should work`, () => {
    const file = tree.get('/test/sales-agents-list/list.facade.ts');
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
});
