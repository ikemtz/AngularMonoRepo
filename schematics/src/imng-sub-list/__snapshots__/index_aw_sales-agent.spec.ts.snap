// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imng-sub-list actions should work 1`] = `
"import { ODataResult, ODataState } from 'imng-kendo-odata';
import { createPayloadAction } from 'imng-ngrx-utils';

export const loadSalesAgentsRequest = createPayloadAction<{ customerId: string, odataState: ODataState; }>(
    '[SalesAgents] Load SalesAgents Request');
export const loadSalesAgentsSuccess = createPayloadAction<{ customerId: string, odataResult: ODataResult<ISalesAgent>; }>(
    '[SalesAgents] Load SalesAgents Success');
export const reloadSalesAgentsRequest = createPayloadAction<string>(
    '[SalesAgents] Reload SalesAgents Request');
export const reloadSalesAgentsSuccess = createPayloadAction<{ customerId: string, odataResult: ODataResult<ISalesAgent>; }>(
    '[SalesAgents] Reload SalesAgents Success');

export const deleteSalesAgentRequest = createPayloadAction<{ customerId: string; salesAgent: ISalesAgent; }>(
    '[SalesAgents] Delete SalesAgent Request');
"
`;

exports[`imng-sub-list api.service should work 1`] = `
"import { Injectable } from '@angular/core';
import { NrsrxBaseApiClientService } from 'imng-nrsrx-client-utils';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env';

@Injectable({
  providedIn: 'root',
})
export class SalesAgentApiService extends NrsrxBaseApiClientService<ISalesAgent> {
  public override url = environment.webApiEndpoints.salesAgents;
  constructor(http: HttpClient) {
    super(http);
  }
}
"
`;

exports[`imng-sub-list effects should work 1`] = `
"import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { createEffect, Actions, ofType, concatLatestFrom } from '@ngrx/effects';
import { ODataService } from 'imng-kendo-odata';
import { handleEffectError } from 'imng-ngrx-utils';
import { map, switchMap } from 'rxjs/operators';

import * as salesAgentActionTypes from './sales-agent.actions';
import { salesAgentSelectors } from './sales-agent.selectors';
import { SalesAgentApiService } from '../sales-agents-list';

import { environment } from '@env';

@Injectable()
export class SalesAgentEffects {
  constructor(
    private readonly actions$: Actions,
    private readonly odataService: ODataService,
    private readonly store: Store,
    private readonly apiService : SalesAgentApiService,
  ) {}

  loadSalesAgentsEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(salesAgentActionTypes.loadSalesAgentsRequest),
      switchMap((action: ReturnType<typeof salesAgentActionTypes.loadSalesAgentsRequest>) => this.odataService
        .fetch<ISalesAgent>(environment.odataEndpoints.salesAgents, action.payload.odataState)
        .pipe(
          map(odataResult => salesAgentActionTypes.loadSalesAgentsSuccess({ customerId: action.payload.customerId, odataResult })),
          handleEffectError(action))));
  });

  reloadSalesAgentsEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(salesAgentActionTypes.reloadSalesAgentsRequest),
      concatLatestFrom((action: ReturnType<typeof salesAgentActionTypes.reloadSalesAgentsRequest>) =>
        this.store.select(salesAgentSelectors.selectODataState$(action.payload))),
      switchMap(([action, odataState]) => this.odataService
        .fetch<ISalesAgent>(environment.odataEndpoints.salesAgents, odataState, { bustCache: true })
        .pipe(
          map(odataResult => salesAgentActionTypes.reloadSalesAgentsSuccess({ customerId: action.payload, odataResult })),
          handleEffectError(action))));
  });

  deleteSalesAgentEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(salesAgentActionTypes.deleteSalesAgentRequest),
      switchMap((action: ReturnType<typeof salesAgentActionTypes.deleteSalesAgentRequest>) =>
        this.apiService.delete(action.payload.salesAgent).pipe(
          map(() => salesAgentActionTypes.reloadSalesAgentsRequest(action.payload.customerId)),
          handleEffectError(action))));
  });
}
"
`;

exports[`imng-sub-list facade should work 1`] = `
"import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { PagerSettings } from '@progress/kendo-angular-grid';
import { IDataDeleteFacade } from 'imng-kendo-data-entry';
import { IKendoODataGridFacade } from 'imng-kendo-grid-odata';
import { applyFilter, ODataResult, ODataState } from 'imng-kendo-odata';
import { BehaviorSubject, Observable, switchMap } from 'rxjs';

import * as salesAgentActionTypes from '../+state/sales-agent.actions';
import { customersFeature } from '../+state/customer.feature';
import { salesAgentSelectors } from '../+state/sales-agent.selectors';

@Injectable()
export class SalesAgentListFacade implements IKendoODataGridFacade<ISalesAgent>, IDataDeleteFacade<ISalesAgent> {
  private _parentGridId = '';
  public parentGrid$ = new BehaviorSubject<string>('');
  public get parentGridId() {
    return this._parentGridId;
  }
  public set parentGridId(value: string) {
    this._parentGridId = value;
    this.parentGrid$.next(value);
  }

  loading$ = this.store.select(customersFeature.selectLoading);
  gridData$: Observable<ODataResult<ISalesAgent>> = this.parentGrid$.pipe(
    switchMap(x =>
      this.store.select(salesAgentSelectors.selectGridData$(x))));
  gridPagerSettings$: Observable<false | PagerSettings> = this.parentGrid$.pipe(
    switchMap(x =>
      this.store.select(salesAgentSelectors.selectGridPagerSettings$(x))));
  gridODataState$: Observable<ODataState> = this.parentGrid$.pipe(
    switchMap(x =>
      this.store.select(salesAgentSelectors.selectODataState$(x))));

  constructor(private readonly store: Store) {}

  public loadEntities(odataState: ODataState): void {
    odataState = applyFilter({ ...odataState },
      { field: SalesAgentProperties.CUSTOMER_ID, operator: 'eq', value: this.parentGridId });
    this.store.dispatch(salesAgentActionTypes.loadSalesAgentsRequest({
      customerId: this.parentGridId, odataState }));
  }

  public reloadEntities(): void {
    this.store.dispatch(salesAgentActionTypes.reloadSalesAgentsRequest(this.parentGridId));
  }

  public deleteExistingEntity(salesAgent: ISalesAgent): void {
    this.store.dispatch(salesAgentActionTypes.deleteSalesAgentRequest({
      customerId: this.parentGridId, salesAgent }));
  }
}
"
`;

exports[`imng-sub-list facade spec should work 1`] = `
"import { NgModule } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { HttpClient } from '@angular/common/http';
import { readFirst } from 'imng-ngrx-utils/testing';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule, Store } from '@ngrx/store';
import { ODataState, createODataPayload, createODataResult, ODataService } from 'imng-kendo-odata';
import { testDeleteCurrentEntity } from 'imng-kendo-data-entry/testing';
import { Observable, of } from 'rxjs';

import * as customerActionTypes from '../+state/customer.actions';
import * as salesAgentActionTypes from '../+state/sales-agent.actions';
import { SalesAgentEffects } from '../+state/sales-agent.effects';
import { customersFeature } from '../+state/customer.feature';
import { SalesAgentListFacade } from './list.facade';
import { CustomerListEffects } from '../+state/customer-list.effects';
import { environment } from '@env';

describe('SalesAgentListFacade', () => {
  let facade: SalesAgentListFacade;
  let store: Store;
  let httpClient: HttpClient;

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  beforeEach(() => {}); //NOSONAR

  describe('used in NgModule', () => {
    beforeEach(() => {
      @NgModule({
        imports: [
          StoreModule.forFeature(customersFeature),
          EffectsModule.forFeature([CustomerEffects, SalesAgentListEffects]),
        ],
        providers: [
          SalesAgentListFacade,
          { provide: HttpClient, useValue: { get: jest.fn(() => of(createODataPayload([createTestSalesAgent()]))) } },
        ],
      })
      class CustomFeatureModule {}

      @NgModule({
        imports: [
          StoreModule.forRoot({}, { runtimeChecks: environment.runtimeChecks }),
          EffectsModule.forRoot([]),
          CustomFeatureModule,
        ],
      })
      class RootModule {}
      TestBed.configureTestingModule({ imports: [RootModule] });

      store = TestBed.inject(Store);
      facade = TestBed.inject(SalesAgentListFacade);
      httpClient = TestBed.inject(HttpClient);
      facade.parentGridId = 'CUSTOMER_ID';
    });

    /**
     * The initially generated facade::loadEntities() returns an array of (1)
     */
    test('loadEntities() should return a list of (1) with loading == false and httpClient.get is invoked', async () => {
      let list = await readFirst(facade.gridData$);
      expect(list?.data.length).toBe(0);
      store.dispatch(customerActionTypes.loadCustomersSuccess(
        createODataResult([
          { ...createTestCustomer(), id: 'CUSTOMER_ID' }])));
      facade.loadEntities({});

      list = await readFirst(facade.gridData$);
      const loading = await readFirst(facade.loading$);
      expect(list?.data.length).toBe(1);
      expect(loading).toBe(false);

      const gridPagerSettings$ = await readFirst(facade.gridPagerSettings$);
      expect(gridPagerSettings$).toBe(false);

      expect(httpClient.get).toHaveBeenCalledTimes(1);
      expect(httpClient.get).toHaveBeenCalledWith(
        \`customers-odata/odata/v1/SalesAgents?$filter=customerId eq 'CUSTOMER_ID'&$count=true\`);

      facade.reloadEntities();
      expect(httpClient.get).toHaveBeenCalledTimes(2);
    });

    test('reloadEntities() should return empty list with loaded == true', async () => {
      let list = await readFirst(facade.gridData$);
      let isloading = await readFirst(facade.loading$);

      const service: { fetch: (endpoint: string, odataState: ODataState) => Observable<unknown>; } = TestBed.inject(ODataService);
      const response = of({ data: [{ id: 'i ‚ù§' }, { id: 'imng' }, { id: 'üíØ' }], total: 3 });
      service.fetch = jest.fn(() => response);

      expect(list?.data.length).toBe(0);
      expect(isloading).toBe(true);
      store.dispatch(customerActionTypes.loadCustomersSuccess(
        createODataResult([
          { ...createTestCustomer(), id: 'CUSTOMER_ID' }])));
      facade.reloadEntities();

      list = await readFirst(facade.gridData$);
      isloading = await readFirst(facade.loading$);

      expect(list?.data.length).toBe(3);
      expect(isloading).toBe(false);
      expect(service.fetch).toHaveBeenCalledTimes(1);
    });

    test('it should get the grid state', async () => {
      const filteringState: ODataState = {
        filter: { logic: 'and', filters: [{ field: 'üí©', operator: 'eq', value: 'üçë' }] },
      };
      let state = await readFirst(facade.gridODataState$);
      expect(state?.count).toBeUndefined();

      store.dispatch(customerActionTypes.loadCustomersSuccess(
        createODataResult([
          { ...createTestCustomer(), id: 'CUSTOMER_ID' }])));

      facade.loadEntities(filteringState);

      state = await readFirst(facade.gridODataState$);
      expect(state).toStrictEqual(filteringState);

      facade.loadEntities({});
      state = await readFirst(facade.gridODataState$);
      expect(state).toMatchSnapshot();
    });

    /**
     * Use \`salesAgentsLoaded\` to manually submit list for state management
     */
    test('gridData$ should return the loaded list; and loaded flag == true', async () => {
      let list = await readFirst(facade.gridData$);
      expect(list?.data.length).toBe(0);
      store.dispatch(customerActionTypes.loadCustomersSuccess(
        createODataResult([
          { ...createTestCustomer(), id: 'CUSTOMER_ID' }])));
      store.dispatch(salesAgentActionTypes.loadSalesAgentsSuccess(
        { customerId: 'CUSTOMER_ID', odataResult: createODataResult([createTestSalesAgent(), createTestSalesAgent()]) }));

      list = await readFirst(facade.gridData$);
      expect(list?.data.length).toBe(2);
    });

    test('it should handle DeleteItem', async () => {
      await testDeleteCurrentEntity(facade, httpClient);
    });
  });
});
"
`;

exports[`imng-sub-list list component should work 1`] = `
"import { Component, ChangeDetectionStrategy, Input } from '@angular/core';
import { KendoODataBasedComponent } from 'imng-kendo-grid-odata';
import { ODataState } from 'imng-kendo-odata';
import { SalesAgentListFacade } from './list.facade';
import { salesAgentGridState } from './list.grid-state';

@Component({
  selector: 'aw-sales-agent-list',
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.scss'],
  providers: [SalesAgentListFacade],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SalesAgentListComponent extends KendoODataBasedComponent<ISalesAgent, SalesAgentListFacade> {
  public readonly props = SalesAgentProperties;
  public currentItem: ISalesAgent | undefined;

  constructor(facade: SalesAgentListFacade) {
    super(facade, salesAgentGridState);
  }

  @Input()
  public set parentGridId(value: string) {
    this.facade.parentGridId = value;
  }

  public deleteItem(item: ISalesAgent): void {
    this.facade.deleteExistingEntity(item);
  }
}
"
`;

exports[`imng-sub-list list component spec should work 1`] = `
"import { CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA } from '@angular/core';
import { waitForAsync, ComponentFixture, TestBed } from '@angular/core/testing';
import { createDataDeleteMockFacade } from 'imng-kendo-data-entry/testing';
import { createODataGridMockFacade } from 'imng-kendo-grid-odata/testing';
import { StoreModule, Store } from '@ngrx/store';

import { SalesAgentListComponent } from './list.component';
import { SalesAgentListFacade } from './list.facade';


describe('SalesAgentListComponent', () => {
  let component: SalesAgentListComponent;
  let fixture: ComponentFixture<SalesAgentListComponent>;
  let listFacade: SalesAgentListFacade;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [SalesAgentListComponent],
      imports: [
        StoreModule.forRoot({})
        StoreModule.forFeature(customersFeature),
      ],
      providers: [
        { provide: SalesAgentListFacade, useValue: createODataGridMockFacade(createDataDeleteMockFacade()) },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SalesAgentListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    listFacade = TestBed.inject(SalesAgentListFacade);
    component.parentGridId = 'CUSTOMER_ID';
  });

  afterAll(() => {
    component.ngOnDestroy();
  });

  test('it should create', () => {
    expect(component).toBeTruthy();
  });

  test('it should handle reload', () => {
    component.reloadEntities();
    expect(listFacade.reloadEntities).toHaveBeenCalledTimes(1);
  });

  test('it should handle DeleteItem', () => {
    const item = createTestSalesAgent();
    component.deleteItem(item);
    expect(listFacade.deleteExistingEntity).toHaveBeenCalledTimes(1);
    expect(listFacade.deleteExistingEntity).toHaveBeenCalledWith(item);
  });
});
"
`;

exports[`imng-sub-list list html should work 1`] = `
"<kendo-grid [height]="375" kendoGridFocusable [imngODataGrid]="this">
  <ng-template kendoGridToolbarTemplate>
    <div class="w-100 d-flex justify-content-between">
      <h4>Sales Agents</h4>
      <imng-kendo-odata-grid-header
        (reloadEntitiesClicked)="reloadEntities()"
        [hideResetFilters]="true"
        [hasHiddenColumns$]="hasHiddenColumns$">
      </imng-kendo-odata-grid-header>
    </div>
  </ng-template>

  <kendo-grid-column [field]="props.ID" [width]="100" title="Id" filter="numeric" [hidden]="false>
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6">Id</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.NAME" [width]="100" title="Name" filter="text" [hidden]="false>
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6">Name</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [field]="props.LOGIN_ID" [width]="100" title="Login Id" filter="text" [hidden]="false>
    <ng-template class="header" kendoGridHeaderTemplate let-column>
      <span class="h6">Login Id</span>
    </ng-template>
  </kendo-grid-column>
  <kendo-grid-column [width]="25">
    <ng-template kendoGridCellTemplate let-dataItem let-rowIndex="rowIndex">
      <button type="button" class="btn btn-sm me-1" (click)="deleteItem(dataItem)" placement="left" tooltip="Delete">
        <span class="k-icon k-i-delete"></span>
      </button>
    </ng-template>
  </kendo-grid-column>

  <kendo-grid-pdf [fileName]="getExportFileName('SalesAgents')" [allPages]="false">
    <kendo-grid-pdf-margin top="1cm" left="1cm" right="1cm" bottom="1cm"></kendo-grid-pdf-margin>
  </kendo-grid-pdf>
  <kendo-grid-excel [fileName]="getExportFileName('SalesAgents')" [fetchData]="excelData"></kendo-grid-excel>
</kendo-grid>
"
`;

exports[`imng-sub-list models should work 1`] = `
"import { PagerSettings } from '@progress/kendo-angular-grid';
import { ODataResult, ODataState } from 'imng-kendo-odata';

export interface IExtCustomer extends ICustomer {
  salesAgentODataState: ODataState;
  salesAgentOData: ODataResult<ISalesAgent>;
  salesAgentPagerSettings: PagerSettings | false;
}
"
`;

exports[`imng-sub-list selectors should work 1`] = `
"import { createSelector } from '@ngrx/store';
import { PagerSettings } from '@progress/kendo-angular-grid';
import { createEmptyODataResult } from 'imng-kendo-odata';
import { getById } from 'imng-ngrx-utils';
import { customersFeature } from './customer.feature';

const selectSalesAgents = (customerId: string) => createSelector(
  customersFeature.selectGridData,
  gridData => getById(gridData, customerId) || {
    salesAgentOData: createEmptyODataResult<ISalesAgent>(),
    salesAgentODataState: {},
    salesAgentPagerSettings: false as (false | PagerSettings),
  },
);

const selectGridData$ = (customerId: string) => createSelector(
  selectSalesAgents(customerId),
  entity => entity.salesAgentOData,
);
const selectODataState$ = (customerId: string) => createSelector(
  selectSalesAgents(customerId),
  entity => entity.salesAgentODataState,
);
const selectGridPagerSettings$ = (customerId: string) => createSelector(
  selectSalesAgents(customerId),
  entity => entity.salesAgentPagerSettings || false,
);

export const salesAgentSelectors = {
  selectSalesAgents,
  selectGridData$,
  selectODataState$,
  selectGridPagerSettings$,
};

"
`;

exports[`imng-sub-list tree files should match 1`] = `
[
  "/test/+state/sales-agent.actions.ts",
  "/test/+state/sales-agent.effects.ts",
  "/test/+state/sales-agent.reducers.ts",
  "/test/+state/sales-agent.selectors.ts",
  "/test/models/ext-customer.ts",
  "/test/sales-agents-list/api.service.ts",
  "/test/sales-agents-list/index.ts",
  "/test/sales-agents-list/list.component.html",
  "/test/sales-agents-list/list.component.scss",
  "/test/sales-agents-list/list.component.spec.ts",
  "/test/sales-agents-list/list.component.ts",
  "/test/sales-agents-list/list.facade.spec.ts",
  "/test/sales-agents-list/list.facade.ts",
  "/test/sales-agents-list/list.grid-state.ts",
]
`;
